service/raft/raft_group_registry.cc

```diff
@@ -408,6 +413,98 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
 
 raft_group_registry::~raft_group_registry() = default;
 
+namespace {
+    class composite_abort_source {
+        abort_source _as;
+        utils::small_vector<abort_source::subscription, 2> _subscriptions;
+    public:
+        void add(abort_source& as) {
+            as.check();
+            auto sub = as.subscribe([this]() noexcept { _as.request_abort(); });
+            assert(sub);
+            _subscriptions.push_back(std::move(*sub));
+        }
+        abort_source& abort_source() noexcept {
+            return _as;
+        }
+    };
+    auto fmt_loc(const seastar::compat::source_location& l) {
+        return fmt::format("{}({}:{}) `{}`", l.file_name(), l.line(), l.column(), l.function_name());
+    };
+}
+
+raft_server_with_timeouts::raft_server_with_timeouts(raft::server& server, raft_group_registry& registry,
+        std::optional<lowres_clock::duration> timeout)
+    : _server(server)
+    , _registry(registry)
+    , _timeout(timeout)
+{
+}
+
+template <std::invocable<abort_source*> Op>
+std::invoke_result_t<Op, abort_source*>
+raft_server_with_timeouts::run_with_timeout(Op&& op, const char* op_name,
+    seastar::abort_source* as, std::optional<raft_timeout> timeout)
+{
+    if (!timeout) {
+        co_await op(as);
+        co_return;
+    }
+    if (!timeout->value) {
+        if (!_timeout) {
+            on_internal_error(rslog, ::format("raft operation [{}], timeout requested at [{}],"
+                                              "but no value for it has been defined",
+                op_name, fmt_loc(timeout->loc)));
+        }
+        timeout->value = _timeout.value();
+    }
+    composite_abort_source composite_as;
+
+    abort_on_expiry<> expiry{lowres_clock::now() + *timeout->value};
+    composite_as.add(expiry.abort_source());
+
+    if (as) {
+        composite_as.add(*as);
+    }
+
+    try {
+        co_return co_await op(&composite_as.abort_source());
```

Comment:
Is the `composite_abort_source` class really necessary? Can't we just do something like:
```cpp
abort_on_expiry<> expiry{lowres_clock::now() + *timeout->value};
auto sub = expiry.abort_source().subscribe([as]() noexcept { if (as) as->request_abort(); });

try {
    co_return co_await op(as ? *as : expiry.abort_source());
```

Reply:
`subscribe([as]() noexcept { if (as) as->request_abort(); });`
this is incorrect, we need to trigger abort no matter if we have `as` or not.

We could subscribe to `as` and trigger `expiry.abort_source()`, but that would be confusing.

It's better to extract the wiring of `abort_source`-s, it can be reused elsewhere.

Reply:
> `subscribe([as]() noexcept { if (as) as->request_abort(); });` this is incorrect, we need to trigger abort no matter if we have `as` or not.

This change
```cpp
co_return co_await op(as ? as : &expiry.abort_source());
```
should ensure that if we don't have `as` and we time out, we will trigger abort.

I found a different problem with my idea though. We cannot distinguish only `expiry.abort_source()` triggering from both `expiry.abort_source()` and `as` triggering. We would have to treat the second case as a timeout, which is different from the current code. The if below would change to
```cpp
if (!expiry.abort_source().abort_requested()) {
    throw;
}
```
I don't think it really matters. Both abort sources triggering seems like an extremely rare case and we don't know what was triggered first anyway. 

> It's better to extract the wiring of `abort_source`-s, it can be reused elsewhere.

Yes, I don't mind it. It just seemed to me like a bit of an overkill for this particular problem.

Reply:
> This change
> 
> ```c++
> co_return co_await op(as ? as : &expiry.abort_source());
> ```
> 
> should ensure that if we don't have `as` and we time out, we will trigger abort.

We can't trigger the `as` from parameters in case of timeout, we don't own it, some unrelated code can depend on it.

If you want to save on `abort_source`-s,  you can pass `expiry.abort_source()` to `co_await op()` and trigger it through `as.subscribe`, but that would be confusing, as I said above. We would need to change the the `if` below
```
if (!expiry.abort_source().abort_requested() || (as && as->abort_requested())) {
    throw;
}
```
With the current structure each `abort_source` has its own responsibility and the `if` is clear.

Reply:
> We can't trigger the `as` from parameters in case of timeout, we don't own it, some unrelated code can depend on it.

Right.

Thank you for the explanation.